server/src/main/java/io/druid/curator/discovery/CuratorDruidNodeDiscoveryProvider.java

```diff
@@ -0,0 +1,327 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.curator.discovery;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Inject;
+import io.druid.concurrent.Execs;
+import io.druid.discovery.DiscoveryDruidNode;
+import io.druid.discovery.DruidNodeDiscovery;
+import io.druid.discovery.DruidNodeDiscoveryProvider;
+import io.druid.guice.ManageLifecycle;
+import io.druid.guice.annotations.Json;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.lifecycle.LifecycleStart;
+import io.druid.java.util.common.lifecycle.LifecycleStop;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.server.initialization.CuratorDiscoveryConfig;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.PathChildrenCache;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheEvent;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheListener;
+import org.apache.curator.utils.ZKPaths;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+
+/**
+ */
+@ManageLifecycle
+public class CuratorDruidNodeDiscoveryProvider extends DruidNodeDiscoveryProvider
+{
+  private static final Logger log = new Logger(CuratorDruidNodeDiscoveryProvider.class);
+
+  private final CuratorFramework curatorFramework;
+  private final CuratorDiscoveryConfig config;
+  private final ObjectMapper jsonMapper;
+
+  private final ExecutorService listenerExecutor;
+
+  private final Map<String, NodeTypeWatcher> nodeTypeWatchers = new ConcurrentHashMap<>();
+
+  private boolean stopped = false;
+
+  @Inject
+  public CuratorDruidNodeDiscoveryProvider(
+      CuratorFramework curatorFramework,
+      CuratorDiscoveryConfig config,
+      @Json ObjectMapper jsonMapper
+  )
+  {
+    // This is single-threaded to ensure that all listener calls are executed precisesly in the oder of add/remove
+    // event occurences.
+    this.listenerExecutor = Execs.singleThreaded("CuratorDruidNodeDiscoveryProvider-ListenerExecutor");
+    this.curatorFramework = curatorFramework;
+    this.config = config;
+    this.jsonMapper = jsonMapper;
+  }
+
+
+
+  @Override
+  public synchronized DruidNodeDiscovery getForNodeType(String nodeType)
+  {
+    if (stopped) {
+      throw new ISE("Provider has been stopped.");
+    }
+
+    NodeTypeWatcher nodeTypeWatcher = nodeTypeWatchers.get(nodeType);
+    if (nodeTypeWatcher == null) {
+      log.info("Creating/Starting NodeTypeWatcher for nodeType [%s].", nodeType);
```

Comment:
Nit: I read "/" as "OR" which seems weird where we are both creating and starting.  I think the message would be better as just "Creating NodeTypeWatcher" or "Creating and starting NodeTypeWatcher"